### PR TITLE
Add head dim support

### PR DESCRIPTION
**Pull request by aflah02**
Addresses #1434
WIP

### CODE REVIEW

File: megatron/model/transformer.py [hunk 1]
```diff
@@ -337,24 +341,24 @@ def __init__(
             )  # how large the total hidden dim for each of K and V is
         else:
             self.num_kv_heads_per_partition = self.num_attention_heads_per_partition
-            self.kv_hidden_size = neox_args.hidden_size
+            self.kv_hidden_size = self.query_hidden_size
 
         if not self.gqa:
```
Comment (aflah02) [inline megatron/model/transformer.py:346]: ColumnParallelLinear maps residual states into one fused QKV tensor:

  input:  hidden_size
  output: [Q | K | V]

  The input remains hidden_size because attention receives residual-stream vectors.

  For MHA, Q, K, and V each have width query_hidden_size:

  output_size = 3 * query_hidden_size

  For GQA/MQA, Q has query_hidden_size, while K and V each have the smaller kv_hidden_size:

  output_size = query_hidden_size + 2 * kv_hidden_size

  Previously, NeoX assumed query_hidden_size == hidden_size, which is no longer necessarily true.

  For Qwen3-0.6B:

  hidden_size       = 1024
  query_hidden_size = 16 * 128 = 2048
  kv_hidden_size    = 8 * 128  = 1024

  fused output = 2048 + 1024 + 1024 = 4096

  The old expression produced only 1024 + 1024 + 1024 = 3072, under-sizing the query projection.

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -887,9 +891,9 @@ def forward(self, hidden_states, attention_mask, layer_past=None):
         # [b, np, sq, hn] --> [sq, b, np, hn]
         context_layer = context_layer.permute(2, 0, 1, 3).contiguous()
 
-        # [sq, b, np, hn] --> [sq, b, hp]
+        # [sq, b, np, hn] --> [sq, b, qp]
```
Comment (aflah02) [inline megatron/model/transformer.py:894]: After attention, the context tensor contains one result per local query head:

  [sq, batch, num_query_heads / TP, head_dim]

  Flattening its final two dimensions therefore produces:

  (num_query_heads / TP) * head_dim
  = query_hidden_size / TP
  = qp

  So the correct reshape is:

  new_context_layer_shape = context_layer.size()[:-2] + (
      self.query_hidden_size_per_partition,
  )

  Previously, hidden_size_per_partition worked only because query_hidden_size == hidden_size.

  For Qwen3-0.6B with TP=2:

  local context width = (16 / 2) * 128 = 1024
  hidden_size / TP    = 1024 / 2 = 512  # incorrect

  Using 512 would attempt to reshape a 1024-wide context incorrectly. After flattening to qp, the
  row-parallel output projection contracts the global query width back to the residual hidden_size.
